From b3b3ab5400c78d4f4de8169b2b5c83cd1957d419 Mon Sep 17 00:00:00 2001
From: Michael Natterer <mitch@gimp.org>
Date: Fri, 23 Mar 2012 16:25:36 +0100
Subject: [PATCH 3/5] gdk: add signal GdkWindow::native-child-event

and emit it whenever we recieve an NSEvent on a subview
of the toplevel, which we can't have added ourselves.
---
 gdk/gdkmarshalers.list        |    1 +
 gdk/gdkwindow.c               |   14 ++++++++++++++
 gdk/quartz/gdkevents-quartz.c |    3 +++
 3 files changed, 18 insertions(+), 0 deletions(-)

diff --git a/gdk/gdkmarshalers.list b/gdk/gdkmarshalers.list
index ea36bae..7b37163 100644
--- a/gdk/gdkmarshalers.list
+++ b/gdk/gdkmarshalers.list
@@ -4,3 +4,4 @@ VOID:POINTER,POINTER,POINTER
 OBJECT:VOID
 OBJECT:DOUBLE,DOUBLE
 VOID:DOUBLE,DOUBLE,POINTER,POINTER
+VOID:POINTER,POINTER
diff --git a/gdk/gdkwindow.c b/gdk/gdkwindow.c
index d20b424..0ab1516 100644
--- a/gdk/gdkwindow.c
+++ b/gdk/gdkwindow.c
@@ -126,6 +126,7 @@ enum {
   PICK_EMBEDDED_CHILD, /* only called if children are embedded */
   TO_EMBEDDER,
   FROM_EMBEDDER,
+  NATIVE_CHILD_EVENT,
   LAST_SIGNAL
 };
 
@@ -585,6 +586,19 @@ gdk_window_class_init (GdkWindowObjectClass *klass)
 		  G_TYPE_DOUBLE,
 		  G_TYPE_POINTER,
 		  G_TYPE_POINTER);
+
+  signals[NATIVE_CHILD_EVENT] =
+    g_signal_new (g_intern_static_string ("native-child-event"),
+		  G_OBJECT_CLASS_TYPE (object_class),
+		  G_SIGNAL_RUN_LAST,
+		  0,
+		  NULL, NULL,
+		  _gdk_marshal_VOID__POINTER_POINTER,
+		  G_TYPE_NONE,
+		  2,
+		  G_TYPE_POINTER,
+		  G_TYPE_POINTER);
+
 }
 
 static void
diff --git a/gdk/quartz/gdkevents-quartz.c b/gdk/quartz/gdkevents-quartz.c
index ece0f07..06dd4d9 100644
--- a/gdk/quartz/gdkevents-quartz.c
+++ b/gdk/quartz/gdkevents-quartz.c
@@ -737,6 +737,9 @@ find_window_for_ns_event (NSEvent *nsevent,
                     if (r.origin.x <= *x && r.origin.x + r.size.width >= *x &&
                         r.origin.y <= *y && r.origin.y + r.size.height >= *y)
                       {
+                        g_signal_emit_by_name (toplevel, "native-child-event",
+                                               sv, nsevent);
+
                         /* event is within subview, forward back to Cocoa */
                         return NULL;
                       }
-- 
1.7.5.4

